### ISSUE TITLE
AMI zope errors related to PYTHON_INSTALL_LAYOUT

### ISSUE THREAD

**Issue by douglarek** [certbot/certbot#2823]
```

       __|  __|_  )
       _|  (     /   Amazon Linux AMI
      ___|\___|___|

https://aws.amazon.com/amazon-linux-ami/2016.03-release-notes/
```

```
./letsencrypt-auto certonly --standalone --email www@icocoa.me -d icocoa.me -d www.icocoa.me --debug
Checking for new version...
Creating virtual environment...
Installing Python packages...
Installation succeeded.
Requesting root privileges to run letsencrypt...
   sudo /home/ec2-user/.local/share/letsencrypt/bin/letsencrypt certonly --standalone --email www@icocoa.me -d icocoa.me -d www.icocoa.me --debug
Traceback (most recent call last):
  File "/home/ec2-user/.local/share/letsencrypt/bin/letsencrypt", line 7, in <module>
    from letsencrypt.main import main
  File "/home/ec2-user/.local/share/letsencrypt/local/lib/python2.7/dist-packages/letsencrypt/main.py", line 11, in <module>
    import zope.component
  File "/home/ec2-user/.local/share/letsencrypt/local/lib/python2.7/dist-packages/zope/component/__init__.py", line 16, in <module>
    from zope.interface import Interface
ImportError: No module named interface
```

```
/home/ec2-user/.local/share/letsencrypt/bin/pip install zope.interface interface
Collecting zope.interface
Collecting interface
Requirement already satisfied (use --upgrade to upgrade): setuptools in /home/ec2-user/.local/share/letsencrypt/lib/python2.7/dist-packages (from zope.interface)
Collecting zope.schema (from interface)
Requirement already satisfied (use --upgrade to upgrade): zope.event in /home/ec2-user/.local/share/letsencrypt/lib/python2.7/dist-packages (from zope.schema->interface)
Installing collected packages: zope.interface, zope.schema, interface
Successfully installed interface zope.interface zope.schema-4.4.2
You are using pip version 8.0.3, however version 8.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

```
/home/ec2-user/.local/share/letsencrypt/bin/python
Python 2.7.10 (default, Dec  8 2015, 18:25:23)
[GCC 4.8.3 20140911 (Red Hat 4.8.3-9)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from zope.interface import Interface
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named interface
```

master branch 9d7519ca25684b81455696ceea2a22bc94ca8862

**Comment by ricog** [comment]
I had similar errors, but was able to get it working with these commands (as root): https://github.com/letsencrypt/letsencrypt/issues/1680#issuecomment-176729326

**Comment by robertsky** [comment]
Having the same issue. The commands that @ricog suggested didn't work. There is a similar report made earlier, #2804.

**Comment by keemyb** [comment]
`export PYTHON_INSTALL_LAYOUT=amzon` as root was what got this working for me, as stated here https://github.com/letsencrypt/letsencrypt/issues/1680#issuecomment-175298203

**Comment by douglarek** [comment]
Having the same issue. The commands that @keemyb  suggested didn't work.

**Comment by keemyb** [comment]
@douglarek what about [this comment](https://github.com/letsencrypt/letsencrypt/issues/1680#issuecomment-176729326)?

**Comment by douglarek** [comment]
@keemyb https://github.com/letsencrypt/letsencrypt/issues/1680#issuecomment-176729326 not work too

**Comment by MadMaxUa** [comment]
unset PYTHON_INSTALL_LAYOUT as root was what got this working for me. i found solution here [http://stackoverflow.com/questions/30983243/django-installed-in-dist-packages](stackoverflow)

**Comment by douglarek** [comment]
@MadMaxUa thanks, `unset PYTHON_INSTALL_LAYOUT` works!

**Comment by douglarek** [comment]
`certbot v0.7.0` is broken now:

```python
Traceback (most recent call last):
  File "/usr/bin/virtualenv", line 9, in <module>
    load_entry_point('virtualenv==12.0.7', 'console_scripts', 'virtualenv')()
  File "/usr/lib/python2.7/dist-packages/virtualenv.py", line 829, in main
    symlink=options.symlink)
  File "/usr/lib/python2.7/dist-packages/virtualenv.py", line 997, in create_environment
    install_wheel(to_install, py_executable, search_dirs)
  File "/usr/lib/python2.7/dist-packages/virtualenv.py", line 965, in install_wheel
    'PIP_NO_INDEX': '1'
  File "/usr/lib/python2.7/dist-packages/virtualenv.py", line 907, in call_subprocess
    % (cmd_desc, proc.returncode))
OSError: Command /home/ec2-user/.loca...ncrypt/bin/python2.7 -c "import sys, pip; sys...d\"] + sys.argv[1:]))" setuptools pip failed with error code 2
```

**Comment by pornel** [comment]
I'm still seeing this problem on AWS EC2 (using git ffc1874854fde7f61870780ce8684af2978b2434), but this issue and #2872  both have been closed.

`unset PYTHON_INSTALL_LAYOUT` helped, but I wouldn't have found that out without trying to file a but about this :)

**Comment by pde** [comment]
Possibly we should mitigate this by printing a warning if PYTHON_INSTALL_LAYOUT is set?  Or by catching the `ImportError` and printing a suggestion in that case?

**Comment by pombredanne** [comment]
IMHO This `PYTHON_INSTALL_LAYOUT` is a rather bad undocumented quirk made by AMZN patching without documenting the standard Python disutils. Someone should ask them to clear up their acts or document this.

**Comment by robertclemens** [comment]
Can someone update me on this?
unset PYTHON_INSTALL_LAYOUT did not help. Maybe I'm missing something?

**Comment by swoodford** [comment]
+1

**Comment by mattwoods11** [comment]
I had the same issue from what appears to be a bad install of certbot. I fixed it by running the following:

unset PYTHON_INSTALL_LAYOUT
/root/.local/share/letsencrypt/bin/pip install --upgrade certbot

**Comment by sakurai-youhei** [comment]
FYI: I needed 2 steps.

```
rm -rf /root/.local/share/letsencrypt/
PYTHON_INSTALL_LAYOUT= certbot-auto certonly --debug ....
```

**certbot-auto --version**
certbot 0.15.0

**uname -a**
Linux ip-172-31-22-179 4.9.32-15.41.amzn1.x86_64 #1 SMP Thu Jun 22 06:20:54 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**set | grep PYTHON**
PYTHON_INSTALL_LAYOUT=amzn

**Comment by ChristopherSchultz** [comment]
@sakurai-youhei This seemed to work for me. Time to add `--no-self-upgrade` to my crontab. :)

**Comment by chenrui333** [comment]
In my case, I took the advice from [here](https://community.letsencrypt.org/t/solved-usr-local-sbin-certbot-auto-851-usr-local-sbin-certbot-auto-opt-eff-org-certbot-venv-bin-letsencrypt-not-found/42545/4):
`rm -rf /opt/eff.org/certbot/`

Then the script `./certbot-auto --version --debug` works for me now (I also did the `pip` and `virtualenv` upgrades):
```
# ./certbot-auto --version --debug
Bootstrapping dependencies for Amazon... (you can skip this with --no-bootstrap)
yum is /usr/bin/yum
Loaded plugins: priorities, update-motd, upgrade-helper
amzn-main                                                                                                                                                | 2.1 kB  00:00:00
amzn-updates                                                                                                                                             | 2.3 kB  00:00:00
93 packages excluded due to repository priority protections
Package gcc-4.8.3-3.20.amzn1.noarch already installed and latest version
Package augeas-libs-1.0.0-5.7.amzn1.x86_64 already installed and latest version
Package 1:openssl-1.0.1k-15.99.amzn1.x86_64 already installed and latest version
Package 1:openssl-devel-1.0.1k-15.99.amzn1.x86_64 already installed and latest version
Package libffi-devel-3.0.13-16.5.amzn1.x86_64 already installed and latest version
Package system-rpm-config-9.0.3-42.28.amzn1.noarch already installed and latest version
Package ca-certificates-2015.2.6-65.0.1.16.amzn1.noarch already installed and latest version
Package python27-2.7.12-2.120.amzn1.x86_64 already installed and latest version
Package python27-devel-2.7.12-2.120.amzn1.x86_64 already installed and latest version
Package python27-virtualenv-12.0.7-1.13.amzn1.noarch already installed and latest version
Package python27-tools-2.7.12-2.120.amzn1.x86_64 already installed and latest version
Package python27-pip-6.1.1-1.23.amzn1.noarch already installed and latest version
Nothing to do
Creating virtual environment...
Installing Python packages...
Installation succeeded.
certbot 0.18.1
```

**Comment by TheNavigat** [comment]
I'll start researching this issue tomorrow. I'll keep this thread updated with findings.

**Comment by TheNavigat** [comment]
Using Amazon Linux official docker image, `certbot-auto` worked perfectly after installing all dependencies. However, `PYTHON_INSTALL_LAYOUT` was not set by default. Setting `PYTHON_INSTALL_LAYOUT` to `amzn` and running `certbot-auto` failed as mentioned above.

Still researching why this happens.

**Comment by TheNavigat** [comment]
I set up 2 different virtual environments where in one environment I've installed the `zope.interface` package with `PYTHON_INSTALL_LAYOUT` unset, and for the second one, I've installed the package with `PYTHON_INSTALL_LAYOUT` set to `amzn`.

Using pip to install the `zope.interface` package using the dependency requirements defined in the `certbot-auto` script prints the following:

```
Collecting zope.interface==4.1.3 (from -r /root/dependency-requirements.txt (line 5))
  Downloading zope.interface-4.1.3.tar.gz (141kB)
    100% |################################| 143kB 315kB/s 
Requirement already satisfied: setuptools in ./testset/lib/python2.7/site-packages (from zope.component==4.2.2->-r /root/dependency-requirements.txt (line 1))
Installing collected packages: zope.interface
  Running setup.py install for zope.interface ... done
Successfully installed zope.interface-4.1.3
```

However, upon running Python on the faulty virtual environment and running the following code:

```
import pip #needed to use the pip functions
for i in pip.get_installed_distributions(local_only=True):
    print(i)
```

Only the following is printed:

```
wheel 0.30.0
setuptools 36.7.2
pip 9.0.1
```

It should be noted that using the Python version inside the virtual environment where `zope.interface` was installed with `PYTHON_INSTALL_LAYOUT` not set, the following output was printed:

```
zope.interface 4.1.3
wheel 0.30.0
setuptools 36.7.2
pip 9.0.1
```

tl;dr: For some reason when `PYTHON_INSTALL_LAYOUT` is set to `amzn`, pip says that `zope.interface` is installed correctly. However, it's never actually installed.

Still investigating further.

**Comment by TheNavigat** [comment]
I ran `pip` in verbose mode to be able to observe what exactly happens when I try to install `zope.interface`. The output when `PYTHON_INSTALL_LAYOUT` was set to `amzn` was different from when it wasn't set. Here are the tailing ends of the logs, where differences can be observed.

With `PYTHON_INSTALL_LAYOUT` set to `amzn`:

```
    copying zope/interface/tests/test_verify.py -> /root/testset/lib64/python2.7/dist-packages/zope/interface/tests
    copying zope/interface/tests/advisory_testing.py -> /root/testset/lib64/python2.7/dist-packages/zope/interface/tests
    copying zope/interface/tests/m2.py -> /root/testset/lib64/python2.7/dist-packages/zope/interface/tests
    copying zope/interface/tests/test_adapter.py -> /root/testset/lib64/python2.7/dist-packages/zope/interface/tests
    running install_egg_info
    removing '/root/testset/lib64/python2.7/dist-packages/zope.interface-4.1.3-py2.7.egg-info' (and everything under it)
    Copying src/zope.interface.egg-info to /root/testset/lib64/python2.7/dist-packages/zope.interface-4.1.3-py2.7.egg-info
    Installing /root/testset/lib64/python2.7/dist-packages/zope.interface-4.1.3-py2.7-nspkg.pth
    running install_scripts
    writing list of installed files to '/tmp/pip-xKPwss-record/install-record.txt'
    Running setup.py install for zope.interface: finished with status 'done'
  Removing source in /tmp/pip-build-hPGjIs/zope.interface
Successfully installed zope.interface-4.1.3
Cleaning up...
```

With `PYTHON_INSTALL_LAYOUT` not set:

```
    copying zope/interface/tests/test_verify.py -> /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests
    copying zope/interface/tests/advisory_testing.py -> /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests
    copying zope/interface/tests/m2.py -> /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests
    copying zope/interface/tests/test_adapter.py -> /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/interfaces.py to interfaces.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/advice.py to advice.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/declarations.py to declarations.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/registry.py to registry.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/__init__.py to __init__.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/exceptions.py to exceptions.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/verify.py to verify.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/_compat.py to _compat.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/ro.py to ro.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/document.py to document.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/adapter.py to adapter.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/_flatten.py to _flatten.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/interface.py to interface.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/interfaces.py to interfaces.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/idatetime.py to idatetime.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/__init__.py to __init__.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/sequence.py to sequence.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/mapping.py to mapping.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/tests/__init__.py to __init__.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/tests/test_import_interfaces.py to test_import_interfaces.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/tests/test_idatetime.py to test_idatetime.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/common/tests/basemapping.py to basemapping.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_element.py to test_element.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_interface.py to test_interface.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/idummy.py to idummy.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_ro.py to test_ro.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_odd_declarations.py to test_odd_declarations.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/dummy.py to dummy.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/__init__.py to __init__.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/m1.py to m1.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_sorting.py to test_sorting.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/ifoo.py to ifoo.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_advice.py to test_advice.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/ifoo_other.py to ifoo_other.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_document.py to test_document.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_registry.py to test_registry.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_exceptions.py to test_exceptions.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_interfaces.py to test_interfaces.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_declarations.py to test_declarations.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/odd.py to odd.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_verify.py to test_verify.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/advisory_testing.py to advisory_testing.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/m2.py to m2.pyc
    byte-compiling /root/testnoset/lib64/python2.7/site-packages/zope/interface/tests/test_adapter.py to test_adapter.pyc
    running install_egg_info
    Copying src/zope.interface.egg-info to /root/testnoset/lib64/python2.7/site-packages/zope.interface-4.1.3-py2.7.egg-info
    Installing /root/testnoset/lib64/python2.7/site-packages/zope.interface-4.1.3-py2.7-nspkg.pth
    running install_scripts
    writing list of installed files to '/tmp/pip-8Dakj3-record/install-record.txt'
    Running setup.py install for zope.interface: finished with status 'done'
  Removing source in /tmp/pip-build-V6AHmP/zope.interface
Successfully installed zope.interface-4.1.3
Cleaning up...
```

**Comment by TheNavigat** [comment]
The behaviour persists in newer versions of the `zope.interface` package. However, no differences in pip installation verbose logs are observed then.

**Comment by TheNavigat** [comment]
The directory structure of both installations differ.

Here's the structure of the correct installation (with `PYTHON_INSTALL_LAYOUT` not set): https://gist.github.com/TheNavigat/865685c729e3428c027cacf8a02be6b2

Here's the directory structure with `PYTHON_INSTALL_LAYOUT=amzn`: https://gist.github.com/TheNavigat/7485706f7d84bb1e2a132c3b0f24f6f9

Diff: https://gist.github.com/TheNavigat/d0e1dd065495e043d8af52b69f8be45a

**Comment by bradsolves** [comment]
@TheNavigat Thank you for picking up this issue and taking it forward. Following the commands suggested by @chenrui333 (among others), I am receiving the following:

> Requesting to rerun ./certbot-auto with root privileges...
Bootstrapping dependencies for Amazon... (you can skip this with --no-bootstrap)
yum is /usr/bin/yum
Loaded plugins: priorities, update-motd, upgrade-helper
amzn-main                                                                                                            | 2.1 kB  00:00:00     
amzn-updates                                                                                                         | 2.5 kB  00:00:00     
Package gcc-4.8.5-1.22.amzn1.noarch already installed and latest version
Package augeas-libs-1.0.0-5.7.amzn1.x86_64 already installed and latest version
Package 1:openssl-1.0.2k-8.105.amzn1.x86_64 already installed and latest version
Package 1:openssl-devel-1.0.2k-8.105.amzn1.x86_64 already installed and latest version
Package libffi-devel-3.0.13-16.5.amzn1.x86_64 already installed and latest version
Package system-rpm-config-9.0.3-42.28.amzn1.noarch already installed and latest version
Package ca-certificates-2015.2.6-65.0.1.16.amzn1.noarch already installed and latest version
Package python27-2.7.12-2.121.amzn1.x86_64 already installed and latest version
Package python27-devel-2.7.12-2.121.amzn1.x86_64 already installed and latest version
Package python27-virtualenv-15.1.0-1.14.amzn1.noarch already installed and latest version
Package python27-tools-2.7.12-2.121.amzn1.x86_64 already installed and latest version
Package python27-pip-9.0.1-1.24.amzn1.noarch already installed and latest version
Nothing to do
Creating virtual environment...
Installing Python packages...
Installation succeeded.
Traceback (most recent call last):
  File "/opt/eff.org/certbot/venv/bin/letsencrypt", line 7, in <module>
    from certbot.main import main
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/dist-packages/certbot/main.py", line 9, in <module>
    import zope.component
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/dist-packages/zope/component/__init__.py", line 16, in <module>
    from zope.interface import Interface
ImportError: No module named interface


not sure if this helps, but I wanted to add to the log. I am happy to try other commands and provide the output as well.

Thank you in advance.

**Comment by TheNavigat** [comment]
@bradsolves Unfortunately this doesn't add much. Unsetting `PYTHON_INSTALL_LAYOUT` fixes the problem, but what I'm investigating is why this happens. The way `pip`'s behaviour changes because of the variable isn't clear. If the solution is to indeed unset the variable, we need to understand why, and whether it breaks anything else.

**Comment by tseaver** [comment]
I don't know from Amazon Linux, but am a maintainer for `zope.interface`.  Is it possible that configuring the `amzn` layout sets up Python such that an older / incompatible version of `zope.interface` is already installed?

One way to test this hypothesis would be to forcibly upgrade it, e.g. `$ pip install -U zope.interface`.

**Comment by TheNavigat** [comment]
@tseaver I used this file, https://github.com/certbot/certbot/blob/master/letsencrypt-auto-source/pieces/dependency-requirements.txt, to install `zope.interface`. The file lists the specific version and hash. It resulted in the error above when `PYTHON_INSTALL_LAYOUT` was set to `amzn`. When the exact same package (`pip install -r dependency-requirements.txt` - I modifed the file to only include the `zope.interface` package) was installed with `PYTHON_INSTALL_LAYOUT` unset, it worked.

I don't think the version is the problem.
